### PR TITLE
py-qiskit: updates

### PR DESCRIPTION
**Pull request by aweits**

### CODE REVIEW

**Review by tldahlgren — DISMISSED** [review]
Confirmed (new) sha256 for each of the 5 packages.

File: var/spack/repos/builtin/packages/py-qiskit-terra/package.py [hunk 1]
```diff
@@ -15,23 +15,33 @@ class PyQiskitTerra(PythonPackage):
...
-    depends_on("py-psutil@5:", type=("build", "run"))
     depends_on("py-sympy@1.3:", type=("build", "run"))
     depends_on("py-dill@0.3:", type=("build", "run"))
-    depends_on("py-fastjsonschema@2.10:", type=("build", "run"))
-    depends_on("py-python-constraint@1.4:", type=("build", "run"))
     depends_on("py-python-dateutil@2.8.0:", type=("build", "run"))
-    depends_on("py-symengine@0.7:", type=("build", "run"))
-    depends_on("py-tweedledum@1.1:1", type=("build", "run"))
+    depends_on("py-tweedledum@1.1:1", type=("build", "run"), when="^python@:3.11")
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-qiskit-terra/package.py]: If you want to simplify this, you can use:
```python
with when("@0.22.2"):
    depends_on("python@3.7:", type=("build", "run"))
    ....

with when("@0.18.3"):
    depends_on("python@3.6:", type=("build", "run"))
    ...
```
Then you don't need to repeat the `when`.

File: var/spack/repos/builtin/packages/py-qiskit-terra/package.py [hunk 2]
```diff
@@ -15,23 +15,33 @@ class PyQiskitTerra(PythonPackage):
...
+    depends_on("python@3.6:", type=("build", "run"), when="@0.18.3")
+    depends_on("py-cython@0.27.1:", type="build", when="@0.18.3")
+    depends_on("py-contextvars@2.4:", when="@0.18.3 ^python@:3.6", type=("build", "run"))
+    depends_on("py-jsonschema@2.6:", type=("build", "run"), when="@0.18.3")
+    depends_on("py-retworkx@0.9.0:", type=("build", "run"), when="@0.18.3")
+    depends_on("py-scipy@1.4:", type=("build", "run"), when="@0.18.3")
+    depends_on("py-psutil@5:", type=("build", "run"), when="@0.18.3")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-qiskit-terra/package.py]:
```suggestion
    depends_on("py-psutil@5:", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-qiskit-terra/package.py [hunk 3]
```diff
@@ -15,23 +15,33 @@ class PyQiskitTerra(PythonPackage):
...
+    depends_on("py-fastjsonschema@2.10:", type=("build", "run"), when="@0.18.3")
+    depends_on("py-python-constraint@1.4:", type=("build", "run"), when="@0.18.3")
+    depends_on("py-symengine@0.7:", type=("build", "run"), when="@0.18.3")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-qiskit-terra/package.py]: Missing dep:
```
stevedore>=3.0.0
```

File: var/spack/repos/builtin/packages/py-shared-memory38/package.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class PySharedMemory38(PythonPackage):
+    """Backport of multiprocessing.shared_memory for Python 3.6 and
+    3.7. Simply import all things from shared_memory to make your code
+    work. Note that multiprocessing.managers.SharedMemoryManager is also
+    included under shared_memory package."""
+
+    homepage = "https://github.com/mars-project/shared_memory38"
+    url = "https://pypi.io/packages/source/s/shared_memory38/shared_memory38-0.1.2.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-shared-memory38/package.py]: https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#id3

File: var/spack/repos/builtin/packages/py-shared-memory38/package.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+    version("0.1.2", sha256="9d4f11fcb08cce059cc31fc8c70292b786a4cf9e411555b37077b7020f5d97e5")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-shared-memory38/package.py:19]: Missing deps:
```python
depends_on("python@3.6:3.7", type=("build", "run"))
depends_on("py-setuptools", type="build")
```

File: var/spack/repos/builtin/packages/py-symengine/package.py [hunk 6]
```diff
@@ -31,6 +32,7 @@ class PySymengine(PythonPackage):
     depends_on("cmake@2.8.12:", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-symengine/package.py:33]: Python dep is now 3.7:3

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-shared-memory38/package.py (hunk 4 above)
Comment (aweits) [reply var/spack/repos/builtin/packages/py-shared-memory38/package.py]: oops, my template script is broken. fixed.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**